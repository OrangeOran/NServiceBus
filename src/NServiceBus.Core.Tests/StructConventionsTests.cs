namespace NServiceBus.Core.Tests
{
    using System;
    using System.Collections.Generic;
    using System.Runtime.InteropServices;
    using System.Text;
    using ApprovalTests;
    using NUnit.Framework;

    [TestFixture]
    public class StructConventionsTests
    {
        [Test]
        public void ApproveStructsWhichDontFollowStructGuidelines()
        {
            var approvalBuilder = new StringBuilder();
            approvalBuilder.AppendLine(@"-------------------------------------------------- REMEMBER --------------------------------------------------
CONSIDER defining a struct instead of a class if instances of the type are small and commonly short-lived or are commonly embedded in other objects.

AVOID defining a struct unless the type has all of the following characteristics:
   * It logically represents a single value, similar to primitive types(int, double, etc.).
   * It has an instance size under 16 bytes.
   * It is immutable.
   * It will not have to be boxed frequently.

In all other cases, you should define your types as classes.
-------------------------------------------------- REMEMBER --------------------------------------------------
");

            var assembly = typeof(Endpoint).Assembly;
            foreach (var type in assembly.GetTypes())
            {
                if (!type.IsValueType || type.IsEnum || type.IsSpecialName) continue;

                var offendedRules = new List<string> { $"{type.FullName} violates the following rules:" };

                InspectSizeOfStruct(type, offendedRules);
                InspectWhetherStructReferenceReferenceTypes(type, offendedRules);

                if (offendedRules.Count <= 1) continue;
                foreach (var offendedRule in offendedRules)
                {
                    approvalBuilder.AppendLine(offendedRule);
                }
                approvalBuilder.AppendLine();
            }

            Approvals.Verify(approvalBuilder.ToString());
        }

        static void InspectWhetherStructReferenceReferenceTypes(Type type, List<string> offendedRules)
        {
            var mutabilityRules = new List<string> { "   - It is mutable. Following members are mutable:" };

            var fields = type.GetFields();
            foreach (var fieldInfo in fields)
            {
                if(fieldInfo.FieldType == typeof(string)) continue;

                if (fieldInfo.FieldType.IsClass || fieldInfo.FieldType.IsInterface)
                {
                    mutabilityRules.Add($"      - Field {fieldInfo.Name} of type { fieldInfo.FieldType } is potentially mutable.");
                }
            }

            var properties = type.GetProperties();
            foreach (var property in properties)
            {
                if (property.PropertyType == typeof(string)) continue;

                if (property.PropertyType.IsClass || property.PropertyType.IsInterface || property.CanWrite)
                {
                    mutabilityRules.Add($"      - Property {property.Name} of type { property.PropertyType } is potentially mutable.");
                }
            }

            if(mutabilityRules.Count > 1)
                offendedRules.AddRange(mutabilityRules);
        }

        static void InspectSizeOfStruct(Type type, List<string> offendedRules)
        {
            try
            {
                var sizeOf = Marshal.SizeOf(type);
                if (IsLargerThanSixteenBytes(sizeOf))
                {
                    offendedRules.Add($"   - With {sizeOf} bytes it is larger than 16 bytes.");
                }
            }
            catch (Exception)
            {
                offendedRules.Add($"   - The size cannot be determined. Probably this type violates all struct rules.");
            }
        }

        static bool IsLargerThanSixteenBytes(int sizeOf)
        {
            return sizeOf > 16;
        }
    }
}