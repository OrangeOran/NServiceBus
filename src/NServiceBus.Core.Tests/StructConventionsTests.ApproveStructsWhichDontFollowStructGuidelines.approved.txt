-------------------------------------------------- REMEMBER --------------------------------------------------
CONSIDER defining a struct instead of a class if instances of the type are small and commonly short-lived or are commonly embedded in other objects.

AVOID defining a struct unless the type has all of the following characteristics:
   * It logically represents a single value, similar to primitive types(int, double, etc.).
   * It has an instance size under 16 bytes.
   * It is immutable.
   * It will not have to be boxed frequently.

In all other cases, you should define your types as classes.
-------------------------------------------------- REMEMBER --------------------------------------------------

NServiceBus.Faults.FailedMessage violates the following rules:
   - The size cannot be determined. Probably this type violates all struct rules.
   - It is mutable. Following members are mutable:
      - Property Headers of type System.Collections.Generic.Dictionary`2[System.String,System.String] is potentially mutable.
      - Property Body of type System.Byte[] is potentially mutable.
      - Property Exception of type System.Exception is potentially mutable.

NServiceBus.Faults.FirstLevelRetry violates the following rules:
   - The size cannot be determined. Probably this type violates all struct rules.
   - It is mutable. Following members are mutable:
      - Property Headers of type System.Collections.Generic.Dictionary`2[System.String,System.String] is potentially mutable.
      - Property Body of type System.Byte[] is potentially mutable.
      - Property Exception of type System.Exception is potentially mutable.

NServiceBus.Faults.SecondLevelRetry violates the following rules:
   - The size cannot be determined. Probably this type violates all struct rules.
   - It is mutable. Following members are mutable:
      - Property Headers of type System.Collections.Generic.Dictionary`2[System.String,System.String] is potentially mutable.
      - Property Body of type System.Byte[] is potentially mutable.
      - Property Exception of type System.Exception is potentially mutable.

NServiceBus.Features.SLAMonitoring+EstimatedTimeToSLABreachCounter+DataPoint violates the following rules:
   - With 24 bytes it is larger than 16 bytes.

